### PR TITLE
Scaffold programmatic deployment of Observation Service

### PR DESCRIPTION
**Pull request by terryyylim**
**What this PR does / why we need it**:

This PR adds support for programmatic helm deployment of Observation Service using `helm.sh/helm/v3/pkg` package.
- https://helm.sh/docs/topics/advanced/#package-overview

In scope:
- Observation Service can be created and upgraded via Dataset Service `POST/PUT /observation_services` endpoints respectively

Not in-scope:
- DB configuration for storing Observation Service deployment metadata

**Which issue(s) this PR fixes**:

`NONE`

### CODE REVIEW

File: dataset-service/appcontext/appcontext.go [hunk 1]
```diff
@@ -14,13 +14,26 @@ type AppContext struct {
...
+	// Init Services
+	var allServices services.Services
+
 	mlpSvc, err := services.NewMLPService(cfg.MLPConfig.URL)
 	if err != nil {
 		return nil, errors.Wrapf(err, "Failed initializing MLP Service")
 	}
 
-	allServices := services.NewServices(
+	obsSvc := services.NewObservationService(
```
Comment (tiopramayudi) [inline dataset-service/appcontext/appcontext.go:25]: why observation service instance creation needs to know about `allServices`? Also to create `allServices` you need `observation service` instance, it looks like that depending on each other for initialization.

File: dataset-service/appcontext/appcontext.go [hunk 2]
```diff
@@ -14,13 +14,26 @@ type AppContext struct {
...
+		&allServices,
+		cfg.DeploymentConfig,
+		cfg.ObservationServiceConfig,
+	)
+	if err != nil {
```
Comment (tiopramayudi) [inline dataset-service/appcontext/appcontext.go:30]: no err returning when calling `NewObservationService`, i this one can be removed?

File: dataset-service/config/config.go [hunk 3]
```diff
@@ -13,10 +13,18 @@ import (
 type Config struct {
 	Port int `envconfig:"PORT" default:"8080"`
 
-	DeploymentConfig common_config.DeploymentConfig
-	MLPConfig        *MLPConfig
-	NewRelicConfig   newrelic.Config
-	SentryConfig     sentry.Config
+	DeploymentConfig         common_config.DeploymentConfig
```
Comment (tiopramayudi) [inline dataset-service/config/config.go:16]: nit: IMO it feels weird using `snack case` in golang, can we use camelCase for this?

File: dataset-service/config/config_test.go [hunk 4]
```diff
@@ -31,6 +31,11 @@ func TestDefaultConfigs(t *testing.T) {
 			IgnoreStatusCodes: []int{},
 			Labels:            emptyInterfaceMap,
 		},
+		ObservationServiceConfig: ObservationServiceConfig{
```
Comment (tiopramayudi) [inline dataset-service/config/config_test.go:34]: nit: it can be `ObbservationServiceConfig{}`

File: dataset-service/config/config_test.go [hunk 5]
```diff
@@ -67,6 +72,11 @@ func TestLoadConfigFiles(t *testing.T) {
 					IgnoreStatusCodes: []int{403, 404, 405},
 					Labels:            map[string]interface{}{"env": "dev"},
 				},
+				ObservationServiceConfig: ObservationServiceConfig{
```
Comment (tiopramayudi) [inline dataset-service/config/config_test.go:75]: nit: it can be ObbservationServiceConfig{}

File: dataset-service/controller/observation_service_controller.go [hunk 6]
```diff
@@ -58,15 +58,19 @@ func (o ObservationServiceController) CreateObservationService(
 	r *timberv1.CreateObservationServiceRequest,
 ) (*timberv1.CreateObservationServiceResponse, error) {
 	// Check if the projectId is valid
-	err := o.checkProject(r.GetProjectId())
+	projectID := r.GetProjectId()
+	project, err := o.appCtx.Services.MLPService.GetProject(projectID)
```
Comment (tiopramayudi) [inline dataset-service/controller/observation_service_controller.go:62]: nit: can we reduce this line by store all services in variable like
```
services := o.appCtx.Services
project, err := services.MLPService.GetProject(projectID)
```
so the next line doesn't need to use `o.appCtx.Services` to access ObservationService instance or can you use different context for this controller so you don't pass whole app context, in this case you only need `ObservationService` and `MlpService` instance in the context of this controller

File: dataset-service/services/observation_service.go [hunk 7]
```diff
@@ -0,0 +1,398 @@
...
+	"github.com/caraml-dev/timber/common/log"
+	timberv1 "github.com/caraml-dev/timber/dataset-service/api"
+	"github.com/caraml-dev/timber/dataset-service/config"
+	"github.com/caraml-dev/timber/dataset-service/models"
+	osConfig "github.com/caraml-dev/timber/observation-service/config"
+)
+
+const (
+	// HELM_DRIVER is the storage backend to be used. Documentation details: https://helm.sh/docs/topics/advanced/#storage-backends
+	HELM_DRIVER = "secret"
```
Comment (tiopramayudi) [inline dataset-service/services/observation_service.go:28]: can we make it private?

File: dataset-service/services/observation_service.go [hunk 8]
```diff
@@ -0,0 +1,398 @@
...
+	// RELEASE_NAME is the helm release name prefix to be used when deploying Observation Service.
+	RELEASE_NAME = "observation-service"
```
Comment (tiopramayudi) [inline dataset-service/services/observation_service.go:30]: can we make it private?

File: dataset-service/services/observation_service.go [hunk 9]
```diff
@@ -0,0 +1,398 @@
...
+	// Environment Variables
+	envVars := []models.Env{
+		{
+			Name:  "GOOGLE_APPLICATION_CREDENTIALS",
+			Value: "/etc/gcp_service_account/service-account.json",
+		},
+	}
+	values.ObservationServiceConfig.ExtraEnvs = envVars
+	// Resources
+	values.ObservationServiceConfig.Resources.Requests.CPU = "1"
```
Comment (tiopramayudi) [inline dataset-service/services/observation_service.go:272]: can we make all of this configurable?

File: dataset-service/services/observation_service.go [hunk 10]
```diff
@@ -0,0 +1,398 @@
...
+	// Compute Fluentd
+	if config.GetSink().GetType() == timberv1.ObservationServiceDataSinkType_OBSERVATION_SERVICE_DATA_SINK_TYPE_FLUENTD {
+		values, err = setFluentdConfig(values, config)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Convert type of values for merging
+	var interfaceValues map[string]interface{}
```
Comment (tiopramayudi) [inline dataset-service/services/observation_service.go:233]: nit: can we use `any` instead of `interface{}`?

File: dataset-service/services/observation_service.go [hunk 11]
```diff
@@ -0,0 +1,398 @@
...
+) ObservationService {
+	return &observationService{
+		services:                 services,
+		gcpProject:               observationServiceConfig.GCPProject,
+		deploymentConfig:         deploymentConfig,
+		observationServiceConfig: observationServiceConfig,
+	}
+}
+
+func (o *observationService) CreateService(
```
Comment (tiopramayudi) [inline dataset-service/services/observation_service.go:67]: `CreateService` and `UpdateService` share a lot of same code, can we extract it out? so no duplication?

File: dataset-service/services/observation_service.go [hunk 12]
```diff
@@ -0,0 +1,398 @@
...
+	projectReleaseName := fmt.Sprintf("%s-%s", RELEASE_NAME, config.GetServiceName())
+
+	// Initialize helm installation
+	installation := action.NewInstall(actionConfig)
+	installation.ReleaseName = projectReleaseName
+	installation.Namespace = caramlProjectName
+	installation.CreateNamespace = true
+
+	// Trigger helm installation
+	release, err := installation.Run(chart, updatedChartValues)
```
Comment (tiopramayudi) [inline dataset-service/services/observation_service.go:107]: does it wait until success?

File: dataset-service/controller/observation_service_controller.go (hunk 6 above)
Comment (terryyylim) [reply dataset-service/controller/observation_service_controller.go]: Passing just `services.Services` now, instead of the entire app context.

File: dataset-service/config/config.go (hunk 3 above)
Comment (terryyylim) [reply dataset-service/config/config.go]: Good point, in fact I think the convention is to simply be lower-cased. I've made it `commonconfig` instead.

File: dataset-service/appcontext/appcontext.go (hunk 1 above)
Comment (terryyylim) [reply dataset-service/appcontext/appcontext.go:25]: Removed!

File: dataset-service/services/observation_service.go (hunk 9 above)
Comment (terryyylim) [reply dataset-service/services/observation_service.go:275]: Do you see this as part of `config.ObservationServiceConfig` and configured as a one-size fit-all solution when deploying Dataset Service?

File: dataset-service/services/observation_service.go (hunk 12 above)
Comment (terryyylim) [reply dataset-service/services/observation_service.go:88]: It doesn't, I think we can use a goroutine to track this status and write the status to DB whenever we integrate the persistence layer to Dataset Service, wdyt?

File: dataset-service/services/observation_service.go (hunk 9 above)
Comment (tiopramayudi) [reply dataset-service/services/observation_service.go:275]: my thought is administrator can configure the default value, but later this value can be override, but for starter we don't need to make it configurable for user right now

File: dataset-service/services/observation_service.go (hunk 12 above)
Comment (tiopramayudi) [reply dataset-service/services/observation_service.go:88]: yes that sounds good, side question if there are 2 requests with the same payload for create observation service, the second request would be fail?

File: dataset-service/controller/observation_service_controller_test.go [hunk 13]
```diff
@@ -25,17 +25,22 @@ func (s *ObservationServiceControllerTestSuite) SetupSuite() {
...
+	projectName := "test-project"
+	expectedProject := &mlp.Project{Id: 0, Name: projectName}
 	mlpSvc.On("GetProject", projectID).Return(expectedProject, nil)
 	mlpSvc.On(
 		"GetProject", int64(3),
 	).Return(nil, errors.Newf(errors.NotFound, "MLP Project info for id %d not found in the cache", int64(3)))
 
+	// Create mock Observation service and set up with test responses
+	observationSvc := &mocks.ObservationService{}
+	observationSvc.On("CreateService", projectName, mock.Anything).Return(&timberv1.ObservationServiceResponse{}, nil)
```
Comment (krithika369) [inline dataset-service/controller/observation_service_controller_test.go:37]: Are we not checking the failure scenarios?

File: dataset-service/models/observation_service_helm.go [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+
+import (
+	"fmt"
+
+	timberv1 "github.com/caraml-dev/timber/dataset-service/api"
+	os "github.com/caraml-dev/timber/observation-service/config"
+)
+
+// GetFluentdConfigModel converts FluentdConfig proto to Observation Service's FluentdConfig
+func GetFluentdConfigModel(config *timberv1.FluentdConfig, projectName string) *os.FluentdConfig {
```
Comment (krithika369) [inline dataset-service/models/observation_service_helm.go:11]: Probably better to call this: `NewFluentdConfig` and `NewKafkaConfig` as they are like constructors. Since they are in the `models` package, the `Model` part of the name may be redundant too.

File: dataset-service/services/observation_service.go [hunk 15]
```diff
@@ -0,0 +1,401 @@
...
+	"helm.sh/helm/v3/pkg/chart"
+	"helm.sh/helm/v3/pkg/chart/loader"
+	"helm.sh/helm/v3/pkg/cli"
+
+	commonconfig "github.com/caraml-dev/timber/common/config"
+	"github.com/caraml-dev/timber/common/log"
+	timberv1 "github.com/caraml-dev/timber/dataset-service/api"
+	"github.com/caraml-dev/timber/dataset-service/config"
+	"github.com/caraml-dev/timber/dataset-service/models"
+	osConfig "github.com/caraml-dev/timber/observation-service/config"
```
Comment (leonlnj) [inline dataset-service/services/observation_service.go:23]: nit: `osConfig` `commonconfig` different case 😅

File: dataset-service/models/observation_service_helm.go [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+
+// Env represents configurable environment parameters for Observation Service deployment
+type Env struct {
+	Name  string `json:"name"`
+	Value string `json:"value"`
+}
+
+// ConfigurableResources represents configurable resource parameters for Observation Service deployment
+type ConfigurableResources struct {
+	CPU    string `json:"cpu"`
```
Comment (krithika369) [inline dataset-service/models/observation_service_helm.go:43]: Shall we use k8s.io/apimachinery/pkg/api/resource `resource.Quantity` for this? That will also help us validate if the provided resource is valid, right? It can be converted to string again before Helm apply.

File: dataset-service/models/observation_service_helm.go [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+	Requests ConfigurableResources `json:"requests"`
+	Limits   ConfigurableResources `json:"limits"`
+}
+
+// Autoscaling represents configurable autoscaling parameters for Observation Service deployment
+type Autoscaling struct {
+	Enabled                        bool `json:"enabled"`
+	MinReplicas                    int  `json:"minReplicas"`
+	MaxReplicas                    int  `json:"maxReplicas"`
+	TargetCPUUtilizationPercentage int  `json:"targetCPUUtilizationPercentage"`
```
Comment (krithika369) [inline dataset-service/models/observation_service_helm.go:58]: Shall we also add `TargetMemoryUtilizationPercentage` ?

File: dataset-service/models/observation_service_helm_test.go [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	timberv1 "github.com/caraml-dev/timber/dataset-service/api"
+	os "github.com/caraml-dev/timber/observation-service/config"
+)
+
+func TestFluendConfigToApiSchema(t *testing.T) {
```
Comment (krithika369) [inline dataset-service/models/observation_service_helm_test.go:13]: Would be better if the test case name closely follows the API name - `TestGetFluentdConfigModel` in this case (or `TestNewFluentdConfig` if adopting the name change suggested [here](https://github.com/caraml-dev/timber/pull/12/files#r1066544424).

File: dataset-service/services/observation_service.go [hunk 19]
```diff
@@ -0,0 +1,401 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	return chart, nil
+}
+
+func isChartInstallable(ch *chart.Chart) (bool, error) {
+	switch ch.Metadata.Type {
+	case "", "application":
```
Comment (leonlnj) [inline dataset-service/services/observation_service.go:198]: for my understanding why these options for the `Metadata.Type` specifically? 

I thought `isChartInstallable` would actually call `chart.Metadata.Validate()` and do other form of validation on the chart to mark it as 'installable'. But I think the validation can be done on publishing of helm chart instead, so probably just a method to check the metadata instead of marking it as installable - or there will be more validation that is required?

File: dataset-service/services/observation_service.go [hunk 20]
```diff
@@ -0,0 +1,401 @@
...
+	// helm_driver is the storage backend to be used. Documentation details: https://helm.sh/docs/topics/advanced/#storage-backends
+	helm_driver = "secret"
+	// release_name is the helm release name prefix to be used when deploying Observation Service.
+	release_name = "observation-service"
+)
+
+// ObservationService provides a set of methods to interact with the MLP APIs
+type ObservationService interface {
+	// CreateService creates new Observation Service Helm release and returns ID of created Observation Service
+	CreateService(projectName string, config *timberv1.ObservationServiceConfig) (*timberv1.ObservationServiceResponse, error)
```
Comment (krithika369) [inline dataset-service/services/observation_service.go:36]: I think it's better to not use the generated server API protos in the service layer and keep it contained to the controller layer - this will keep the service layer shielded from changes to the API contracts. Instead, we can either create a new set of service data models (though it may seem redundant at first, it can help as the code evolves), or, alternatively, use the `models` package types (the controller will do the conversion before calling the service API).

A similar comment about the return type - using `*timberv1.ObservationServiceResponse` seems unintuitive for service methods. It's preferable they don't know about the concept of request / response.

Let me know if you have differing views.

File: dataset-service/services/observation_service.go [hunk 21]
```diff
@@ -0,0 +1,401 @@
...
+	commonconfig "github.com/caraml-dev/timber/common/config"
+	"github.com/caraml-dev/timber/common/log"
+	timberv1 "github.com/caraml-dev/timber/dataset-service/api"
+	"github.com/caraml-dev/timber/dataset-service/config"
+	"github.com/caraml-dev/timber/dataset-service/models"
+	osConfig "github.com/caraml-dev/timber/observation-service/config"
+)
+
+const (
+	// helm_driver is the storage backend to be used. Documentation details: https://helm.sh/docs/topics/advanced/#storage-backends
```
Comment (krithika369) [inline dataset-service/services/observation_service.go:27]: Curious - What's the behavior when we choose secret? Is a new secret created per release or per namespace? It's not very clear from the docs.

File: dataset-service/services/observation_service.go (hunk 12 above)
Comment (krithika369) [reply dataset-service/services/observation_service.go:88]: I think we can track the installation state changes in the DB and block an attempt to deploy an already "pending deployment" using that. Wdyt?

File: dataset-service/services/observation_service.go [hunk 22]
```diff
@@ -0,0 +1,401 @@
...
+	}
+
+	// Postfix helm release name with provided Service name as a CaraML project could have multiple different services
+	projectReleaseName := fmt.Sprintf("%s-%s", release_name, config.GetServiceName())
+
+	// Initialize helm installation
+	installation := action.NewInstall(actionConfig)
+	installation.ReleaseName = projectReleaseName
+	installation.Namespace = caramlProjectName
+	installation.CreateNamespace = true
```
Comment (krithika369) [inline dataset-service/services/observation_service.go:85]: I imagine Helm will need some special RBAC to be able to create namespaces. If so, shall we document that in the README ?

File: dataset-service/services/observation_service.go [hunk 23]
```diff
@@ -0,0 +1,401 @@
...
+
+	// Trigger helm installation
+	release, err := installation.Run(chart, updatedChartValues)
+	if err != nil {
+		log.Error(err)
+		return nil, err
+	}
+	log.Infof("%s helm release installation (version %d) underway! Status: %s", release.Name, release.Version, release.Info.Status)
+	// Pretty print Helm release YAML manifest with fmt.Println
+	fmt.Println(release.Manifest)
```
Comment (krithika369) [inline dataset-service/services/observation_service.go:95]: Why `fmt.Println` here? Shall we use `log.Infof` here too, so we can turn it off if not needed (or `log.Debugf`)?

File: dataset-service/services/observation_service.go [hunk 24]
```diff
@@ -0,0 +1,401 @@
...
+	chart, err := loader.LoadDir(chartDir)
+	if err != nil {
+		return nil, err
+	}
+
+	return chart, nil
+}
+
+func isChartInstallable(ch *chart.Chart) (bool, error) {
+	switch ch.Metadata.Type {
```
Comment (krithika369) [inline dataset-service/services/observation_service.go:197]: Why do we need this check? What is it checking?

File: dataset-service/services/observation_service.go [hunk 25]
```diff
@@ -0,0 +1,401 @@
...
+
+	return chart, updatedChartValues, actionConfig, nil
+}
+
+func readChart() (*chart.Chart, error) {
+	// TODO: Publish helm chart and reference published chart
+	// Locate directory of Observation Service helm chart
+	_, filename, _, _ := runtime.Caller(0)
+	repositoryRoot := filepath.Dir(filepath.Dir(path.Dir(filename)))
+	chartDir := fmt.Sprintf("%s/infra/charts/observation-service", repositoryRoot)
```
Comment (leonlnj) [inline dataset-service/services/observation_service.go:187]: does this get updated? If so where is it done? 

If there was an upgrade (v2)  after the initial (v1), will sequent update read from (v2) and apply the changes?

File: dataset-service/services/observation_service.go [hunk 26]
```diff
@@ -0,0 +1,401 @@
...
+		return nil, nil, nil, err
+	}
+
+	// Retrieve computed chart values based on default values and request body values
+	updatedChartValues, err := getChartValues(config, gcpProject, caramlProjectName, deploymentConfig, observationServiceConfig)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	// Generate configuration required to run helm upgrade
```
Comment (leonlnj) [inline dataset-service/services/observation_service.go:171]: nit: update doc to indicate both helm install and upgrade

File: dataset-service/services/observation_service.go [hunk 27]
```diff
@@ -0,0 +1,401 @@
...
+		},
+	}
+	values.ObservationServiceConfig.ExtraEnvs = envVars
+	// Resources
+	values.ObservationServiceConfig.Resources.Requests.CPU = "1"
+	values.ObservationServiceConfig.Resources.Requests.Memory = "512Mi"
+	values.ObservationServiceConfig.Resources.Limits.CPU = "1"
+	values.ObservationServiceConfig.Resources.Limits.Memory = "1Gi"
+	// Autoscaling
+	values.ObservationServiceConfig.Autoscaling.Enabled = false
```
Comment (krithika369) [inline dataset-service/services/observation_service.go:280]: Are we not enabling autoscaling by default?

File: dataset-service/services/observation_service.go [hunk 28]
```diff
@@ -0,0 +1,401 @@
...
+
+	// --- Fluentd configs --- //
+	// Image
+	values.FluentdConfig.Image.Tag = observationServiceConfig.FluentdImageTag
+	// Environment Variables
+	envVars = []models.Env{
+		{
+			Name:  "FLUENTD_WORKER_COUNT",
+			Value: "1",
```
Comment (krithika369) [inline dataset-service/services/observation_service.go:289]: Are these the defaults? I thought that based on previous tests, we were going to use a higher worker count.

File: dataset-service/services/observation_service.go [hunk 29]
```diff
@@ -0,0 +1,401 @@
...
+			Name:  "FLUENTD_GCP_PROJECT",
+			Value: gcpProject,
+		},
+		{
+			Name:  "FLUENTD_BQ_DATASET",
+			Value: caramlProjectName,
+		},
+		{
+			Name:  "FLUENTD_BQ_TABLE",
+			Value: fmt.Sprintf("%s_observation_log", caramlProjectName),
```
Comment (krithika369) [inline dataset-service/services/observation_service.go:325]: There is also a dataset name that must be factored in ?

File: dataset-service/services/observation_service.go (hunk 12 above)
Comment (tiopramayudi) [reply dataset-service/services/observation_service.go:88]: we can do that once integrate the db. I just want to understand for current state (no db yet) the same payload to create observation service, the second request would be fail, since the chart already installed, is that correct?

File: dataset-service/services/observation_service.go [hunk 30]
```diff
@@ -0,0 +1,401 @@
...
+func getChartValues(
+	config *timberv1.ObservationServiceConfig,
+	gcpProject string,
+	caramlProjectName string,
+	deploymentConfig commonconfig.DeploymentConfig,
+	observationServiceConfig config.ObservationServiceConfig,
+) (map[string]any, error) {
+	// Initialize and set default values
+	values := &models.ObservationServiceHelmValues{}
+	values = setDefaultValues(values, gcpProject, caramlProjectName, config.GetServiceName(), deploymentConfig, observationServiceConfig)
```
Comment (leonlnj) [inline dataset-service/services/observation_service.go:213]: similar concern as the `readChart()`, i would expect default to be used during `install` only, or when the chart does not exist. 

Although might not be a problem now, it might possibly introduce side effect of reverting something that is not set explicitly down this method.

File: dataset-service/services/observation_service.go [hunk 31]
```diff
@@ -0,0 +1,401 @@
...
+	values.ObservationServiceConfig.Resources.Requests.CPU = "1"
+	values.ObservationServiceConfig.Resources.Requests.Memory = "512Mi"
+	values.ObservationServiceConfig.Resources.Limits.CPU = "1"
+	values.ObservationServiceConfig.Resources.Limits.Memory = "1Gi"
+	// Autoscaling
+	values.ObservationServiceConfig.Autoscaling.Enabled = false
+
+	// --- Fluentd configs --- //
+	// Image
+	values.FluentdConfig.Image.Tag = observationServiceConfig.FluentdImageTag
```
Comment (krithika369) [inline dataset-service/services/observation_service.go:284]: Just realised that right now, the `fluentd` sink type is tied to BQ: https://github.com/caraml-dev/timber/blob/main/dataset-service/proto/caraml/timber/v1/observation_service.proto#L34

Whereas, it's possible have fluentd publish logs to kafka too. How does the observation service know it needs to configure fluentd for BQ ?

File: dataset-service/services/observation_service.go (hunk 23 above)
Comment (terryyylim) [reply dataset-service/services/observation_service.go:95]: As described in the comment above this line, `fmt.Println` is required to pretty print it in YAML format. Using  `log.Infof` will result in multiple lines with explicit `\n` linebreaks printed.

File: dataset-service/services/observation_service.go (hunk 26 above)
Comment (terryyylim) [reply dataset-service/services/observation_service.go:171]: Do you mean document this in another file?

File: dataset-service/services/observation_service.go (hunk 27 above)
Comment (terryyylim) [reply dataset-service/services/observation_service.go:280]: Are we? We can if we want, I was thinking more in terms of having the eventing mechanism update this config instead, wdyt?

File: dataset-service/services/observation_service.go (hunk 29 above)
Comment (terryyylim) [reply dataset-service/services/observation_service.go:325]: Do you mean BQ dataset name as part of this value or?

File: dataset-service/controller/logwriter_controller.go [hunk 32]
```diff
@@ -5,17 +5,17 @@ import (
...
 	"github.com/caraml-dev/timber/common/log"
 	timberv1 "github.com/caraml-dev/timber/dataset-service/api"
-	"github.com/caraml-dev/timber/dataset-service/appcontext"
+	"github.com/caraml-dev/timber/dataset-service/services"
 )
 
 // LogWriterController implements controller logic for Dataset Service log writer endpoints
 type LogWriterController struct {
-	appCtx *appcontext.AppContext
+	services *services.Services
```
[the remark this answers is not in the document]
Comment (terryyylim) [reply dataset-service/controller/logwriter_controller.go]: Yes that's correct. I changed it because of Tio's comment stating no need to include the entire appContext for now. But it makes sense this will be used in the future anyway so I've reverted this change.

File: dataset-service/services/observation_service.go (hunk 19 above)
Comment (terryyylim) [reply dataset-service/services/observation_service.go:198]: `Metadata.Type` can either be `application` or `library`. We shouldn't be deploying `library` charts. But again, we're only concerned with Observation Service helm chart which will be an `application` chart anyway, this check is redundant, I'll remove it.

File: dataset-service/services/observation_service.go (hunk 24 above)
Comment (terryyylim) [reply dataset-service/services/observation_service.go:197]: `Metadata.Type` can either be `application` or `library`. We shouldn't be deploying `library` charts. But again, we're only concerned with Observation Service helm chart which will be an `application` chart anyway, this check is redundant, I'll remove it.

File: dataset-service/services/observation_service.go (hunk 30 above)
Comment (terryyylim) [reply dataset-service/services/observation_service.go:198]: I agree, but this portion will be combined with existing values once we have them stored in a persistence layer. This function will definitely be updated when a DB has been introduced to Dataset Service.

File: dataset-service/services/observation_service.go (hunk 21 above)
Comment (terryyylim) [reply dataset-service/services/observation_service.go:27]:
```
Helm 3 changed the default release information storage to Secrets in the namespace of the release. Helm 2 by default stores release information as ConfigMaps in the namespace of the Tiller instance.
```
We would require `secret` since the chart uses Helm 3, and if other storage backends are chosen, the release information cannot be retrieved, e.g if we run `helm list -A`, we will not be able to retrieve any helm installations even if there are any. I've tested this behaviour.

File: dataset-service/services/observation_service.go (hunk 12 above)
Comment (terryyylim) [reply dataset-service/services/observation_service.go:88]: >if there are 2 requests with the same payload for create observation service, the second request would be fail?

Yes, it should fail. Ideally, this function should first check the DB if such a helm installation exists and if it does, it should throw an error.

File: dataset-service/services/observation_service.go (hunk 22 above)
Comment (terryyylim) [reply dataset-service/services/observation_service.go:85]: Good question, I wonder if this is already available in a new K8s cluster by default. I actually tried this installation in both Minikube and K3d locally and both worked fine without any intervention to create an RBAC.

File: dataset-service/services/observation_service.go (hunk 31 above)
Comment (terryyylim) [reply dataset-service/services/observation_service.go:269]: That would be via this config - https://github.com/caraml-dev/timber/blob/main/observation-service/config/config.go#L113.

File: dataset-service/services/observation_service.go (hunk 26 above)
Comment (leonlnj) [reply dataset-service/services/observation_service.go:164]: no, just indicating `run helm install/upgrade` since they are not used the same scenario but this config works for both

File: dataset-service/services/observation_service.go (hunk 25 above)
Comment (leonlnj) [reply dataset-service/services/observation_service.go:180]: just understood that this will ideally read from a published chart

Comment (terryyylim) [reply dataset-service/services/observation_service.go:180]: This should ideally be a published helm chart.

File: dataset-service/services/observation_service.go (hunk 20 above)
Comment (terryyylim) [reply dataset-service/services/observation_service.go]: I agree! I've took your suggestion and added response structs to `models` package.

File: dataset-service/services/observation_service.go (hunk 29 above)
Comment (krithika369) [reply dataset-service/services/observation_service.go:310]: Oops. I meant to say "service name" as opposed to "dataset name".

File: dataset-service/services/observation_service.go (hunk 9 above)
Comment (terryyylim) [reply dataset-service/services/observation_service.go]: >why it doesn't seem right? it is more like resource request that user specifies during merlin model /turing router deployment right?

Yes, but what is still relatively unknown here is who are the expected users to Dataset Service? Would it be administrators of CaraML platform? Or service users themselves? Whoever is expected to use it will most likely result in us deciding on different configuration parameters to expose.

A deeper discussion would probably be required to answer this.

tldr; These hardcoded values are likely to be changed (More defaults added/some defaults removed) when we get to implementing the persistence layer.

File: dataset-service/controller/logwriter_controller.go (hunk 32 above)
[the remark this answers is not in the document]
Comment (tiopramayudi) [reply dataset-service/controller/logwriter_controller.go]: yeah, this is from this [comment](https://github.com/caraml-dev/timber/pull/12#discussion_r1065348027). The idea is to reduce the `train wrecks` when trying to accessing mlpService/observationService instance. The simplest one is by changing
```
services := o.appCtx.Services
project, err := services.MLPService.GetProject(projectID)
....
result, err := services.ObservationService.CreateService(project.Name, r.GetObservationService())
....
```
This is due to we pass all the appCtx to the controller, where in fact we don't need the appCtx directly. IMO it is better to have local context for specific controller so it can't access any instance that doesn't relate to that controller

File: dataset-service/controller/observation_service_controller_test.go (hunk 13 above)
Comment (terryyylim) [reply dataset-service/controller/observation_service_controller_test.go]: Added the failure scenarios 🙏🏻

File: dataset-service/services/observation_service.go (hunk 29 above)
Comment (terryyylim) [reply dataset-service/services/observation_service.go:310]: That's a good catch, I think it should be just service name instead of project name, because project name should be the GCP dataset's name.

File: dataset-service/services/observation_service.go (hunk 28 above)
Comment (terryyylim) [reply dataset-service/services/observation_service.go:274]: Similar response in previous thread applicable here - https://github.com/caraml-dev/timber/pull/12#discussion_r1066626402. This is to be resolved in a subsequent PR.

File: dataset-service/services/observation_service.go (hunk 23 above)
Comment (terryyylim) [reply dataset-service/services/observation_service.go]: Makes sense, added a comment and also, added condition to check for `debug` mode prior to printing.

File: dataset-service/models/observation_service_helm_test.go [hunk 33]
```diff
@@ -0,0 +1,47 @@
...
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	timberv1 "github.com/caraml-dev/timber/dataset-service/api"
+	os "github.com/caraml-dev/timber/observation-service/config"
+)
+
+func TestNewFluentdConfigToApiSchema(t *testing.T) {
```
Comment (krithika369) [inline dataset-service/models/observation_service_helm_test.go:13]: What does `ToApiSchema` in this name mean? https://github.com/caraml-dev/timber/pull/12#discussion_r1066549237

File: dataset-service/services/observation_service.go (hunk 20 above)
Comment (krithika369) [reply dataset-service/services/observation_service.go]: The latest changes don't really address my comment here - maybe I wasn't clear.

(You can refer to the XP management service for a similar example: https://github.com/krithika369/turing-experiments/blob/main/management-service/services/experiment_service.go#L79)

The stuff in the `github.com/caraml-dev/timber/dataset-service/api` package are generated from the protos that are closely tied to the API request and response contract. It's better the `service` and `models` layers are completely shielded from them. So, ideally, these services should work with data structures in their own package or deeper (i.e., models). Currently, the input to these service functions is still relying on the proto which would be nice to do away with.

Coming to the return value `*models.ObservationServiceResponse`:

>  It's preferable they don't know about the concept of request / response.

I feel it's more natural for the service to return the model (like gorm DB model) or some other struct local to the services package and not something that is named `...Response` - the response construction is the job of the controller and it can pick up whatever it needs from the values returned by the service method. Since we don't have the DB models yet, may be we can just return `string` Id for now and remove the `ObservationServiceResponse` ?

Hope that makes sense.

File: dataset-service/services/observation_service.go (hunk 27 above)
Comment (krithika369) [reply dataset-service/services/observation_service.go:271]: I think the autoscaling may actually be one of the default values. 🤔 Anyway we can revisit this once the dataset svc workflows are clearer. Could you help add a //TODO note here as a reminder?

File: dataset-service/services/observation_service_test.go [hunk 34]
```diff
@@ -0,0 +1,139 @@
...
+
+func (s *ObservationServiceTestSuite) SetupSuite() {
+	s.Suite.T().Log("Setting up ObservationServiceTestSuite")
+}
+
+func (s *ObservationServiceTestSuite) TearDownSuite() {
+	s.Suite.T().Log("Cleaning up ObservationServiceTestSuite")
+}
+
+func TestObservationService(t *testing.T) {
```
Comment (krithika369) [inline dataset-service/services/observation_service_test.go:29]: For the create / update methods, it may be nice to create some dummy interfaces and use the helm dependencies via dependency injection, so that the actual install can be skipped. May be it's not too important now but could be useful when we start implementing the DB portion of the workflow, so that can be verified. Could you leave a TODO for this in the observation_service file?

File: dataset-service/models/observation_service_helm_test.go (hunk 33 above)
Comment (terryyylim) [reply dataset-service/models/observation_service_helm_test.go]: Was a typo - fixed!

**Review by krithika369 — APPROVED** [review]
LGTM on my end. Thanks for the changes, @terryyylim ! 🎉
